pkg/storj/metainfo.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package storj
+
+import (
+	"context"
+)
+
+// Metainfo represents a database for storing meta-info about objects
+type Metainfo interface {
+	// MetainfoLimits returns limits for this metainfo database
+	Limits() (MetainfoLimits, error)
+
+	// CreateBucket creates a new bucket with the specified information
+	// Database automatically sets different values in the information
+	CreateBucket(ctx context.Context, bucket string, info *Bucket) (Bucket, error)
+	// DeleteBucket deletes bucket
+	DeleteBucket(ctx context.Context, bucket string) error
+	// GetBucket gets bucket information
+	GetBucket(ctx context.Context, bucket string) (Bucket, error)
+
+	// GetObject returns information about an object
+	GetObject(ctx context.Context, bucket string, path Path) (Object, error)
+	// GetObjectStream returns interface for reading the object stream
+	GetObjectStream(ctx context.Context, bucket string, path Path) (ReadonlyStream, error)
+
+	// CreateObject creates an uploading object and returns an interface for uploading Object information
+	CreateObject(ctx context.Context, bucket string, path Path, info Object) (MutableObject, error)
```

Comment:
If the `info` param in `CreateBucket` is a pointer, should it be a pointer in `CreateObject` and `ModifyObject`?

However, it's not quite clean at the moment. What will happen if the caller sets `info.Bucket` that is different than the `bucket` param? Or `info.Version`, `info.Created`, `info.Modified`, `info.IsPrefix`. Will they be respected? Overwritten? An error returned?

It would be better if the `info` param is a supertype of `Object`, i.e. `SettableObjectInfo` (or a better name) that contains only the fields that can be validly set for this call.

Reply:
I spent time thinking about this. I'll currently add comments that it will overwrite and ignore the params, take a pointer and then add a TODO note for reviewing once we have refactored.

The deciding factor for me was that by introducing multiple interfaces/objects it increases the number of converters and types that need to be juggled. It's more convenient to have just one type.

Reply:
My point is that currently the Object type has 15 fields (including those from the embedded Stream type). The caller will be actually allowed to set just a handful of them. It's quite confusing to know which of them can be set and which not.

A type like `SettableObjectInfo` can be embedded in the Object type, just like the Stream type. Go is pretty good at embedding types. What converters exactly would be necessary in this case?

Reply:
We need to start converting them between protobuf types. But, sure... I'll add a `CreateObject` type.